### PR TITLE
PIM-5641: Xlsx writers can write into several files

### PR DESCRIPTION
**Pull request by willy-ahva**
As Mary, I would like to define a limit number of lines for the Excel files in order to manage them easily in Excel.
The parameter "Number of lines / file" can be defined in the export profile in all export profiles in Excel (the limit is 1 000 000 = Excel limit)
If this limit is reached, more than 1 file is created.
By default the limit is 10 000.
The info tip is "Define the limit number of lines per file"
If more than one file, a ZIP is created and a number will be added at the end of the filename (file_1, file_2...). The headers (columns) in the file are the same in all files.

TECHNICAL DETAILS to help the reviewing process :
 - I added a counter on the FlatItemBuffer to know how much lines we have to write. It helps me to know how much file I need to write and simplify Writers code.
    -> it results the `FlatItemBuffer` now implements `\Countable` interface. We can discuss about this implementation that is not required.
 
| Q                                 | A 
| --------------------------------- | ---
| Added Specs                       | done
| Added Behats                      | doing
| Changelog updated                 | no
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

### CODE REVIEW

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 1]
```diff
@@ -82,23 +87,57 @@ public function write(array $items)
      */
     public function flush()
     {
-        $writer = WriterFactory::create(Type::XLSX);
+        $filesNb = 0;
+        $buffer  = $this->flatRowBuffer->getBuffer();
+        $buffer->rewind();
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Is it needed ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 2]
```diff
@@ -82,23 +87,57 @@ public function write(array $items)
...
+
+        // TODO rename the first file "file_1.xlsx" instead of "file.xlsx" if $filesNb > 1
+    }
+
+    protected function forceFileName($filesNb)
+    {
+        if (2 === $filesNb) {
+            $ext = strrchr($this->filePath, '.');
+            $filePath = strstr($this->filePath, $ext, true);
+            $this->filePath = $filePath . '%fileNb%' . $ext;
```
Comment (SamirBoulil) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: sprintf ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 3]
```diff
@@ -82,23 +87,57 @@ public function write(array $items)
...
+
+            $writer = WriterFactory::create(Type::XLSX);
+            $this->fillFile($writer, $buffer);
+            $this->writtenFiles[$this->getPath()] = basename($this->getPath());
+        }
+
+        // TODO rename the first file "file_1.xlsx" instead of "file.xlsx" if $filesNb > 1
+    }
+
+    protected function forceFileName($filesNb)
```
Comment (SamirBoulil) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Can you comment this one, I don't really understand what it is used for

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 1 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Yes, it is. Before, the foreach did it. With the while, you have to do it.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 4]
```diff
@@ -82,23 +87,57 @@ public function write(array $items)
...
-        $writer = WriterFactory::create(Type::XLSX);
+        $filesNb = 0;
+        $buffer  = $this->flatRowBuffer->getBuffer();
+        $buffer->rewind();
+
+        while ($buffer->valid()) {
+            $filesNb++;
+            $this->forceFileName($filesNb);
+
+            $writer = WriterFactory::create(Type::XLSX);
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: As I understand it you instantiate a writer for each line read.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 2 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: `$filePath . '%fileNb%' . $ext` is more readable than `sprintf('%s\%fileNb\%%s', $filePath, $ext)` IMHO.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 3 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: I'll comment later, but this method have to manage the name of the file. If there is only one file, the file has to be name `file.xslx`. If there is more files, they have to be named `file_x.xlsx`. I use the option resolver to do it. I add a token in the file name, and then each time we move to another file, I increment the option.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Nope, for each file because the writer keep in cache the cursor it wrote with. It results an error "stream id idNumber not found" if I don't do that. So I have to instantiate a new Writer for each file.

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php [hunk 5]
```diff
@@ -12,20 +12,24 @@
  * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-class FlatItemBuffer
+class FlatItemBuffer implements \Countable
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: Not sure it's needed

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 6]
```diff
@@ -82,23 +91,77 @@ public function write(array $items)
...
         $hollowItem = array_fill_keys($headers, '');
-        $writer->addRow($headers);
-        foreach ($this->flatRowBuffer->getBuffer() as $incompleteItem) {
+
+        $fileCount = 0;
+        $writtenLinesCount = 0;
+        foreach ($this->flatRowBuffer->getBuffer() as $count => $incompleteItem) {
+            if (0 === $writtenLinesCount % $this->getLinesPerFiles()) {
+                if ($this->severalFilesNeeded()) {
+                    $this->updateFilePathResolverOptions(['fileNb' => sprintf('_%d', ++$fileCount)]);
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Should be incremented just after the close imo.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 7]
```diff
@@ -82,23 +91,77 @@ public function write(array $items)
...
+
+        return $fileCountNeeded > 1;
+    }
+
+    /**
+     * Add a %fileNb% pattern in the file name
+     */
+    protected function enforceFileName()
+    {
+        $ext = strrchr($this->filePath, '.');
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Maybe easier to use [pathinfo](http://php.net/manual/fr/function.pathinfo.php).

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 8]
```diff
@@ -82,23 +91,77 @@ public function write(array $items)
...
+
+    /**
+     * @param array $headers
+     *
+     * @return WriterInterface
+     */
+    protected function getWriter()
+    {
+        $writer = WriterFactory::create(Type::XLSX);
+        $writer->openToFile($this->getPath());
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: I would pass the path as param.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 9]
```diff
@@ -82,23 +91,77 @@ public function write(array $items)
...
+        $totalLineCount  = $hasHeader ? ++$itemCount : $itemCount;
+        $fileCountNeeded = ceil($totalLineCount / $this->getLinesPerFiles());
+
+        return $fileCountNeeded > 1;
+    }
+
+    /**
+     * Add a %fileNb% pattern in the file name
+     */
+    protected function enforceFileName()
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: What about "update" instead of "enforce" ?

Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: I like the `enforce` as it's exactly what we do here. We broke the user's choice to add our patapouille. Maybe others opinion on this naming ?

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php [hunk 10]
```diff
@@ -25,7 +25,8 @@ public function getArchives(JobExecution $jobExecution)
         $directory = dirname($this->getRelativeArchivePath($jobExecution));
         $archives = [];
 
-        foreach ($this->filesystem->listFiles($directory) as $key) {
+        $files = $this->filesystem->listFiles($directory);
```
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php]: what's the purpose of this change???
revert please

File: src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php [hunk 11]
```diff
@@ -145,4 +188,51 @@ public function getWrittenFiles()
...
+     */
+    public function getLinesPerFiles()
+    {
+        return $this->linesPerFiles;
+    }
+
+    /**
+     * @param int $linesPerFiles
+     */
+    public function setLinesPerFiles($linesPerFiles)
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php]: fluent setters ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 12]
```diff
@@ -158,8 +216,37 @@ public function setConfiguration(array $config)
...
+    {
+        return $this->flatRowBuffer->count() > $this->getLinesPerFiles();
+    }
+
+    /**
+     * Add a %fileNb% pattern in the file name
+     */
+    protected function enforceFileName()
+    {
+        $extension = sprintf('.%s', pathinfo($this->filePath, PATHINFO_EXTENSION));
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: pass filepath to the method as argument please
no useless coupling between methods

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 13]
```diff
@@ -158,8 +216,37 @@ public function setConfiguration(array $config)
...
+        $filePath  = strstr($this->filePath, $extension, true);
+        $this->setFilePath($filePath . '%fileNb%' . $extension);
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: return the new filePath
and set it outside

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php [hunk 14]
```diff
@@ -49,22 +57,45 @@ public function write(array $items)
      */
     public function flush()
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: XlsxSimpleWriter and XlsxVariantGroupWriter seem to be very close, no way to decorate ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 15]
```diff
@@ -158,8 +216,37 @@ public function setConfiguration(array $config)
...
-        foreach ($config['mainContext'] as $key => $value) {
-            $this->filePathResolverOptions['parameters']['%' . $key . '%'] = $value;
-        }
+        $this->updateFilePathResolverOptions($config['mainContext']);
+    }
+
+    /**
+     * @return bool
+     */
+    protected function severalFilesNeeded()
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: isSeveralFilesNeeded

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php [hunk 16]
```diff
@@ -42,10 +46,21 @@ public function write(array $items, $addHeader)
...
 
             $this->buffer->write($item);
+            $this->count++;
         }
     }
 
     /**
+     * Return the number of written elements
+     *
+     * @return int
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: inherit doc ?

File: src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php [hunk 17]
```diff
@@ -81,23 +90,46 @@ public function write(array $items)
      */
     public function flush()
     {
-        $writer = WriterFactory::create(Type::XLSX);
-        $writer->openToFile($this->getPath());
+        if ($this->severalFilesNeeded()) {
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php]: it's very hard to understand, not way to split it to make it more understandable ?

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php [hunk 18]
```diff
@@ -104,4 +146,51 @@ public function setWithHeader($withHeader)
...
+     */
+    public function getLinesPerFiles()
+    {
+        return $this->linesPerFiles;
+    }
+
+    /**
+     * @param int $linesPerFiles
+     */
+    public function setLinesPerFiles($linesPerFiles)
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: fluent setter ?

File: spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php [hunk 19]
```diff
@@ -45,6 +45,7 @@ function it_writes_item_to_the_buffer($buffer)
         ], true);
 
         $this->getHeaders()->shouldReturn(['id', 'family']);
+        $this->count()->shouldReturn(2);
```
Comment (phaseinducer) [inline spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php]: It's a different feature, please write it in another test.

**Comment by solivier** [comment]
what about the quick export ?

File: features/Context/catalog/footwear/jobs.yml [hunk 20]
```diff
@@ -322,6 +325,7 @@ jobs:
         type:      export
         configuration:
             withHeader: true
+            linesPerFiles: 10000
```
Comment (solivier) [inline features/Context/catalog/footwear/jobs.yml]: why do you set the linesPerFiles here as it's by default in the writer ? and why 10000 ?

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php (hunk 5 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: done irl

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 21]
```diff
@@ -82,23 +89,46 @@ public function write(array $items)
...
+                $this->writtenFiles[$this->getPath()] = basename($this->getPath());
+                $writtenLinesCount = 0;
+                $fileCount++;
+            }
         }
 
-        $writer->close();
-        $this->writtenFiles[$this->getPath()] = basename($this->getPath());
+        if ($this->severalFilesNeeded()) {
+            $this->setFilePath($originalFilePath);
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: add comment to explain

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 22]
```diff
@@ -82,23 +89,46 @@ public function write(array $items)
...
             $writer->addRow($item);
+            $writtenLinesCount++;
 
             if (null !== $this->stepExecution) {
                 $this->stepExecution->incrementSummaryInfo('write');
             }
+
+            if (0 === $writtenLinesCount % $this->getLinesPerFiles() || $this->flatRowBuffer->count() === $count + 1) {
+                $writer->close();
+                $this->writtenFiles[$this->getPath()] = basename($this->getPath());
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: never use getPath()

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 21 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: remove this patapouille when getPath is removed

File: features/Context/catalog/footwear/jobs.yml (hunk 20 above)
Comment (willy-ahva) [inline features/Context/catalog/footwear/jobs.yml]: What I did in conf of the job is : if you put nothing in the field validation will fail and put 10 000 by default. So I put 10 000 here to initialize at 10 000. And it's a number define by PO.

File: spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php (hunk 19 above)
Comment (willy-ahva) [inline spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php]: Why create another feature exactly the same to just add this line ?

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php (hunk 10 above)
Comment (willy-ahva) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php]: Mistake

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php (hunk 18 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: don't understand

Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: use a fluent setter here

File: spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php (hunk 19 above)
Comment (a2xchip) [inline spec/Pim/Component/Connector/Writer/File/FlatItemBufferSpec.php]: different:
* writes to the buffer
* it implements countable
* it counts headers

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php (hunk 14 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: We wanted to avoid another abstraction layer. Moreover, I have no time to do this in may card. And there is the `setConfiguation()` which is not exactly the same and should be extracted.
In fact I asked myself a long time if I should or not do it and decided to not. But I agree, it should be done.

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php (hunk 18 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: Ok I understood. But no special need here IMHO

File: src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php (hunk 11 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php]: same

**Comment by willy-ahva** [comment]
@solivier good catch for quick export ! I add a default value at 10000 in fixtures.

File: spec/Pim/Component/Connector/Writer/File/XlsxProductWriterSpec.php [hunk 23]
```diff
@@ -119,12 +103,68 @@ function it_prepares_the_export($flatRowBuffer, $mediaCopier, StepExecution $ste
...
 
         $this->flush();
     }
 
+    function it_has_configuration()
+    {
+        $this->getConfigurationFields()->shouldReturnConfiguration();
+    }
+
+    public function getMatchers()
```
Comment (phaseinducer) [inline spec/Pim/Component/Connector/Writer/File/XlsxProductWriterSpec.php]: Can you explain?

File: spec/Pim/Component/Connector/Writer/File/XlsxSimpleWriterSpec.php [hunk 24]
```diff
@@ -83,20 +88,54 @@ function it_writes_the_xlsx_file($flatRowBuffer, BufferInterface $buffer)
...
-                'options' => [
-                    'label' => 'pim_connector.export.withHeader.label',
-                    'help'  => 'pim_connector.export.withHeader.help'
-                ]
-            ],
-        ]);
+        $this->getConfigurationFields()->shouldReturnConfiguration();
+    }
+
+    public function getMatchers()
```
Comment (phaseinducer) [inline spec/Pim/Component/Connector/Writer/File/XlsxSimpleWriterSpec.php]: idem

File: spec/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriterSpec.php [hunk 25]
```diff
@@ -150,9 +134,64 @@ function it_prepares_the_export($flatRowBuffer, $mediaCopier, StepExecution $ste
...
 
         $this->flush();
     }
+
+    function it_has_configuration()
+    {
+        $this->getConfigurationFields()->shouldReturnConfiguration();
+    }
+
+    public function getMatchers()
```
Comment (phaseinducer) [inline spec/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriterSpec.php]: idem

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php [hunk 26]
```diff
@@ -23,7 +23,7 @@
     public function getArchives(JobExecution $jobExecution)
     {
         $directory = dirname($this->getRelativeArchivePath($jobExecution));
-        $archives = [];
+        $archives  = [];
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/AbstractFilesystemArchiver.php]: You didn't modify the file, you can revert imo

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php [hunk 27]
```diff
@@ -87,4 +71,44 @@ public function supports(JobExecution $jobExecution)
...
+
+        if ($writer instanceof ArchivableWriterInterface) {
+            foreach ($writer->getWrittenFiles() as $filePath => $fileName) {
+                if (!is_file($filePath)) {
+                    return false;
+                }
+            }
+        }
+
+        return ($isDeprecatedWriter || $isNewWriter) && is_file($writer->getPath());
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php]: `is_file($writer->getPath())` should not work

File: src/Pim/Component/Connector/Writer/File/AbstractFileWriter.php [hunk 28]
```diff
@@ -82,11 +78,7 @@ public function getFilePath()
      */
     public function getPath()
     {
-        if (null === $this->resolvedFilePath) {
-            $this->resolvedFilePath = $this->filePathResolver->resolve($this->filePath, $this->filePathResolverOptions);
-        }
-
-        return $this->resolvedFilePath;
+        return $this->filePathResolver->resolve($this->filePath, $this->filePathResolverOptions);
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/AbstractFileWriter.php]: :heart:

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 29]
```diff
@@ -22,26 +24,35 @@ class XlsxProductWriter extends AbstractFileWriter implements ItemWriterInterfac
     protected $flatRowBuffer;
 
     /** @var BulkFileExporter */
-    protected $mediaCopier;
+    protected $fileExporter;
 
     /** @var array */
     protected $writtenFiles;
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Move it in last position please

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 30]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+     * Return the given file path with with %fileNb% placeholder just before the extension of the file
+     * ie: in -> '/path/myFile.txt' ; out -> '/path/myFile%fileNb%.txt'
+     *
+     * @param string $fullFilePath
+     *
+     * @return string
+     */
+    protected function getCustomFilePath($fullFilePath)
+    {
+        $extension = sprintf('.%s', pathinfo($fullFilePath, PATHINFO_EXTENSION));
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: `'.' . pathinfo($fullFilePath, PATHINFO_EXTENSION)` is simpler ;)

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 31]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+                    ['parameters' => ['%fileNb%' => sprintf('_%d', $currentFileCount)]]
+                )
+            );
+        }
+
+        return $filePath;
+    }
+
+    /**
+     * Return the given file path with with %fileNb% placeholder just before the extension of the file
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: with x2

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 32]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+     * ie: in -> '/path/myFile.txt' ; out -> '/path/myFile%fileNb%.txt'
+     *
+     * @param string $fullFilePath
+     *
+     * @return string
+     */
+    protected function getCustomFilePath($fullFilePath)
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Bad naming. Maybe `getSuffixedPath()` ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 33]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+    /**
+     * @return bool
+     */
+    protected function isSeveralFilesNeeded()
+    {
+        return $this->flatRowBuffer->count() > $this->getLinesPerFiles();
+    }
+
+    /**
+     * Return the given file path in terms of the current file count if needed
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: I don't understand this sentence.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 34]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+     * @return string
+     */
+    protected function getCurrentFilePath($filePath, $currentFileCount)
+    {
+        if ($this->isSeveralFilesNeeded()) {
+            $filePath = $this->filePathResolver->resolve(
+                $filePath,
+                array_merge_recursive(
+                    $this->filePathResolverOptions,
+                    ['parameters' => ['%fileNb%' => sprintf('_%d', $currentFileCount)]]
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: `sprintf` is useless here

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 35]
```diff
@@ -82,23 +93,40 @@ public function write(array $items)
      */
     public function flush()
     {
-        $writer = WriterFactory::create(Type::XLSX);
-        $writer->openToFile($this->getPath());
+        $filePath = $this->getPath();
+        if ($this->isSeveralFilesNeeded()) {
+            $filePath = $this->getCustomFilePath($this->getPath());
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Naming problem here. I'd say it's rather a `$pathPattern`. What do you think?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 36]
```diff
@@ -82,23 +93,40 @@ public function write(array $items)
...
+        $headers    = $this->flatRowBuffer->getHeaders();
         $hollowItem = array_fill_keys($headers, '');
-        $writer->addRow($headers);
-        foreach ($this->flatRowBuffer->getBuffer() as $incompleteItem) {
+
+        $fileCount = 1;
+        $writtenLinesCount = 0;
+        foreach ($this->flatRowBuffer->getBuffer() as $count => $incompleteItem) {
+            if (0 === $writtenLinesCount % $this->getLinesPerFiles()) {
+                $currentFilePath = $this->getCurrentFilePath($filePath, $fileCount);
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: And here it would be `$filePath = $this->resolveFilePath($pathPattern, $fileCount);`

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 37]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
...
+
+    /**
+     * Return the given file path in terms of the current file count if needed
+     *
+     * @param string $filePath
+     * @param int    $currentFileCount
+     *
+     * @return string
+     */
+    protected function getCurrentFilePath($filePath, $currentFileCount)
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: `resolveFilePath`

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 38]
```diff
@@ -162,4 +217,64 @@ public function setConfiguration(array $config)
             $this->filePathResolverOptions['parameters']['%' . $key . '%'] = $value;
         }
     }
+
+    /**
+     * @return bool
+     */
+    protected function isSeveralFilesNeeded()
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: *are

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 33 above)
Comment (willy-ahva) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: any suggestion ?

File: features/Pim/Behat/Context/Domain/Spread/XlsxFileContext.php [hunk 39]
```diff
@@ -14,11 +15,12 @@ class XlsxFileContext extends PimContext
      * @param string    $code
      * @param TableNode $expectedLines
      *
-     * @Then /^exported xlsx file of "([^"]*)" should contain:$/
+     * @Then /^exported xlsx file( \d+)? of "([^"]*)" should contain:$/
      */
-    public function exportedXlsxFileOfShouldContain($code, TableNode $expectedLines)
+    public function exportedXlsxFileOfShouldContain($number = null, $code, $expectedLines)
```
Comment (pierallard) [inline features/Pim/Behat/Context/Domain/Spread/XlsxFileContext.php]: I don't understand naming if $number variable. Too abstract

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php [hunk 40]
```diff
@@ -87,4 +71,49 @@ public function supports(JobExecution $jobExecution)
...
+     * @return bool
+     */
+    protected function isUsableWriter(ItemWriterInterface $writer)
+    {
+        $isDeprecatedWriter = ($writer instanceof FileWriter);
+        $isNewWriter = ($writer instanceof AbstractFileWriter);
+
+        if (!($isDeprecatedWriter || $isNewWriter)) {
+            return false;
+        }
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php]: you should squash these lines
```if (!($writer instanceof FileWriter) && !($writer instanceof AbstractFileWriter)) {
  return false
}```

File: features/Context/catalog/apparel/jobs.yml [hunk 41]
```diff
@@ -56,7 +56,9 @@ jobs:
         label:     XLSX product export for tablet
         type:      export
         configuration:
+            filePath: /tmp/products.xlsx
             channel:    tablet
+            linesPerFiles: 10000
```
Comment (pierallard) [inline features/Context/catalog/apparel/jobs.yml]: Please update DataGeneratorBundle

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 42]
```diff
@@ -21,27 +23,36 @@ class XlsxProductWriter extends AbstractFileWriter implements ItemWriterInterfac
...
     protected $flatRowBuffer;
 
-    /** @var BulkFileExporter */
-    protected $mediaCopier;
-
     /** @var array */
     protected $writtenFiles;
 
+    /** @var int */
+    protected $linesPerFiles;
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: linesPerFile please

File: features/Pim/Behat/Context/Domain/Spread/XlsxFileContext.php (hunk 39 above)
Comment (phaseinducer) [inline features/Pim/Behat/Context/Domain/Spread/XlsxFileContext.php]: Why remove `TableNode`?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 33 above)
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: "Returns the path of the file to write." ?

(don't understand the sentencer neither)

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php [hunk 43]
```diff
@@ -20,15 +22,26 @@ class XlsxSimpleWriter extends AbstractFileWriter
     /** @var FlatItemBuffer */
     protected $flatRowBuffer;
 
+    /** @var int */
+    protected $linesPerFiles;
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: same

File: src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php (hunk 18 above)
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/XlsxSimpleWriter.php]: setLinesPerFile

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 44]
```diff
@@ -82,23 +93,40 @@ public function write(array $items)
...
                 $this->stepExecution->incrementSummaryInfo('write');
             }
-        }
 
-        $writer->close();
-        $this->writtenFiles[$this->getPath()] = basename($this->getPath());
+            if (0 === $writtenLinesCount % $this->getLinesPerFiles() || $this->flatRowBuffer->count() === $count + 1) {
+                $writer->close();
+                $this->writtenFiles[$filePath] = basename($filePath);
+                $writtenLinesCount = 0;
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: Already reset at line 110.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 33 above)
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: "Return the file path including file count if needed" ?

File: src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php [hunk 45]
```diff
@@ -145,4 +185,81 @@ public function getWrittenFiles()
...
+
+    /**
+     * Return the given file path with %fileNb% placeholder just before the extension of the file
+     * ie: in -> '/path/myFile.txt' ; out -> '/path/myFile%fileNb%.txt'
+     *
+     * @param string $fullFilePath
+     *
+     * @return string
+     */
+    protected function getCustomFilePath($fullFilePath)
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php]: Already said in a previous and deleted comment : bad naming imo, it's confusing and too generic.
What about `getSuffixedFilePath`, `getNumberedFilePath` or `updateFilePath` ?

Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/File/XlsxVariantGroupWriter.php]: Also `$fullFilePath` doesn't mean anything. What about `$originalFilePath` or even just `$filePath` ?

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php (hunk 40 above)
Comment (skeleton) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php]: no, it's more readable to extract this line in 2 variables.

**Comment by pierallard** [comment]
Small comments then GTM

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php (hunk 40 above)
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/FileWriterArchiver.php]: Nope, it was like that before and was split to add semantic.

**Comment by phaseinducer** [comment]
Nothing in the changelog ?

**Comment by momoss** [comment]
GTM
